### PR TITLE
DOC: add type BinaryIO to path param #35505

### LINKED ISSUE #35505: DOC: ExcelWriter constructor types
#### Location of the documentation

[https://pandas.pydata.org/docs/dev/reference/api/pandas.ExcelWriter.html?highlight=excelwriter](https://pandas.pydata.org/docs/dev/reference/api/pandas.ExcelWriter.html?highlight=excelwriter)
or
[https://github.com/pandas-dev/pandas/blob/d9fff2792bf16178d4e450fe7384244e50635733/pandas/io/excel/_base.py#L544](https://github.com/pandas-dev/pandas/blob/d9fff2792bf16178d4e450fe7384244e50635733/pandas/io/excel/_base.py#L544)

#### Documentation problem

Parameter `path` has type `str`, but based on this feature [https://github.com/pandas-dev/pandas/issues/7074](url) I can pass `io.BytesIO`.

#### Suggested fix for documentation

Change type description to `str | io.BytesIO`.

### PR DESCRIPTION
**Pull request by alexhtn**
- [ ] closes #35505
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/io/excel/_base.py [hunk 1]
```diff
@@ -541,7 +541,7 @@ class ExcelWriter(metaclass=abc.ABCMeta):
 
     Parameters
     ----------
-    path : str
+    path : str or BinaryIO
```
Comment (TomAugspurger) [inline pandas/io/excel/_base.py:544]: `io.BytesIO`? What do we actually need here?

Comment (alexhtn) [reply pandas/io/excel/_base.py:544]: `io.Bytes` doesn't fully cover possibilities of ExcelWriter, as I found `BinaryIO` from typeshed is best option, Pycharm doesn't show any warnings

Comment (jreback) [inline pandas/io/excel/_base.py:544]: we should rename this to `path_or_buf` like all other IO routines, in theory need a deprecation for this.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_base.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/io/excel/_base.py:544]: You should have `typing.BinaryIO` here then.

Comment (alexhtn) [reply pandas/io/excel/_base.py]: Done

Comment (alexhtn) [reply pandas/io/excel/_base.py]: Are you expect do this in current pull request?

**Comment by jbrockmendel** [comment]
Seems fine.  Not sure if the extra part of the docstring is necessary.

File: pandas/io/excel/_base.py (hunk 1 above)
Comment (jreback) [reply pandas/io/excel/_base.py]: no this can be done separately (pls make an issue if there isn't one, or if you can just do a PR)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @alexhtn (comment on a follonw)

**Pull request merged**